File: 01jupyter/nb/source/cs02/include/omp_for.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
int main(int argc, char ** argv) {
  char * nthreads_ = getenv("OMP_NUM_THREADS");
  int    nthreads  = (nthreads_ ? atoi(nthreads_) : 1);
  if (nthreads % 32) {
    fprintf(stderr, "OMP_NUM_THREADS (%d) must be a multiple of 32\n", nthreads);
    exit(1);
  }
  int m = (argc > 1 ? atoi(argv[1]) : 5);
  int n = (argc > 2 ? atoi(argv[2]) : 6);
  printf("hello on host\n");
#pragma omp target teams
  {
    printf("in teams: %03d/%03d\n", omp_get_team_num(), omp_get_num_teams());
#pragma omp distribute
    for (int i = 0; i < m; i++) {
      printf("in distribute: i=%03d executed by %03d/%03d\n",
             i, omp_get_team_num(), omp_get_num_teams());
#pragma omp parallel num_threads(nthreads)
      printf("in parallel: i=%03d %03d/%03d %03d/%03d\n",
             i, omp_get_team_num(), omp_get_num_teams(),
             omp_get_thread_num(), omp_get_num_threads());
#pragma omp for
      for (int j = 0; j < n; j++) {
        printf("in for: i=%03d j=%03d executed by %03d/%03d %03d/%03d\n",
               i, j, omp_get_team_num(), omp_get_num_teams(),
               omp_get_thread_num(), omp_get_num_threads());
      }
    }
  }
  printf("back on host\n");
  return 0;
}

File: 01jupyter/nb/source/08_gpu_teams/include/omp_for.f90
program omp_for
  use omp_lib
  integer :: m, n, i, j, nthreads, stat
  character(len=32) :: arg, env
  nthreads = 1
  call get_environment_variable("OMP_NUM_THREADS", env, status=stat)
  if (stat == 0) read (env, *) nthreads
  if (mod(nthreads, 32) /= 0) then
     write (0, "(a,i0,a)") "OMP_NUM_THREADS (", nthreads, ") must be a multiple of 32"
     stop 1
  end if
  m = 5
  n = 6
  if (command_argument_count() > 0) then
     call get_command_argument(1, arg)
     read (arg, *) m
  end if
  if (command_argument_count() > 1) then
     call get_command_argument(2, arg)
     read (arg, *) n
  end if
  print "(a)", "hello on host"
  !$omp target teams
  print "(a,i3.3,a,i3.3)", "in teams: ", omp_get_team_num(), "/", omp_get_num_teams()
  !$omp distribute private(i, j)
  do i = 0, m - 1
     print "(a,i3.3,a,i3.3,a,i3.3)", "in distribute: i=", i, &
          " executed by ", omp_get_team_num(), "/", omp_get_num_teams()
     !$omp parallel num_threads(nthreads) private(j)
     print "(a,i3.3,a,i3.3,a,i3.3,a,i3.3,a,i3.3)", "in parallel: i=", i, " ", &
          omp_get_team_num(), "/", omp_get_num_teams(), " ", &
          omp_get_thread_num(), "/", omp_get_num_threads()
     !$omp do
     do j = 0, n - 1
        print "(a,i3.3,a,i3.3,a,i3.3,a,i3.3,a,i3.3,a,i3.3)", "in for: i=", i, " j=", j, &
             " ", omp_get_team_num(), "/", omp_get_num_teams(), " ", &
             omp_get_thread_num(), "/", omp_get_num_threads()
     end do
     !$omp end do
     !$omp end parallel
  end do
  !$omp end distribute
  !$omp end teams
  print "(a)", "back on host"
end program omp_for
